Bug in BarrageMessageProducer.SnapshotControl.usePreviousValues causes us to always use previous

```
heduler-Concurrent-2 | i.d.s.s.SessionService    | Internal Error '12dfe072-4725-44f9-8dff-422240326a49' io.deephaven.UncheckedDeephavenException: Failed to execute function concurrently despite consistent state
	at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1320)
	at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1157)
	at io.deephaven.engine.table.impl.remote.ConstructSnapshot.constructBackplaneSnapshotInPositionSpace(ConstructSnapshot.java:708)
	at io.deephaven.server.barrage.BarrageMessageProducer.getSnapshot(BarrageMessageProducer.java:2215)
	at io.deephaven.server.barrage.BarrageMessageProducer.updateSubscriptionsSnapshotAndPropagate(BarrageMessageProducer.java:1323)
	at io.deephaven.server.barrage.BarrageMessageProducer$UpdatePropagationJob.run(BarrageMessageProducer.java:999)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.deephaven.server.runner.scheduler.SchedulerModule$ThreadFactory.lambda$newThread$0(SchedulerModule.java:78)
	at org.jpy.PyLib.callAndReturnObject(Native Method)
	at org.jpy.PyObject.call(PyObject.java:449)
	at io.deephaven.server.console.python.DebuggingInitializer.lambda$createInitializer$0(DebuggingInitializer.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: fillChunk precondition broken, invalid range. requested=[300, 319] (or subset), available=[0, 279]
	at io.deephaven.engine.table.impl.sources.ring.AbstractRingChunkSource.fillChunk(AbstractRingChunkSource.java:311)
	at io.deephaven.engine.table.impl.sources.ring.RingColumnSource.fillPrevChunk(RingColumnSource.java:190)
	at io.deephaven.engine.table.impl.sources.LongAsTimeSource.fillPrevChunk(LongAsTimeSource.java:100)
	at io.deephaven.engine.table.impl.sources.UnionColumnSource$FillContext.fillPrevChunkAppend(UnionColumnSource.java:385)
	at io.deephaven.engine.table.impl.sources.UnionColumnSource.fillPrevChunkFromMultipleSources(UnionColumnSource.java:534)
	at io.deephaven.engine.table.impl.sources.UnionColumnSource.fillPrevChunk(UnionColumnSource.java:520)
	at io.deephaven.engine.table.impl.sources.RedirectedColumnSource$FillContext.doOrderedFillAndPermute(RedirectedColumnSource.java:782)
	at io.deephaven.engine.table.impl.sources.RedirectedColumnSource.doFillChunk(RedirectedColumnSource.java:525)
	at io.deephaven.engine.table.impl.sources.RedirectedColumnSource.fillPrevChunk(RedirectedColumnSource.java:504)
	at io.deephaven.engine.table.impl.sources.RedirectedColumnSource$FillContext.doOrderedFillAscending(RedirectedColumnSource.java:805)
	at io.deephaven.engine.table.impl.sources.RedirectedColumnSource.doFillChunk(RedirectedColumnSource.java:520)
	at io.deephaven.engine.table.impl.sources.RedirectedColumnSource.fillPrevChunk(RedirectedColumnSource.java:504)
	at io.deephaven.engine.table.impl.sources.UnboxedTimeBackedColumnSource.fillPrevChunk(UnboxedTimeBackedColumnSource.java:98)
	at io.deephaven.engine.table.impl.remote.ConstructSnapshot.getSnapshotDataAsChunkList(ConstructSnapshot.java:1622)
	at io.deephaven.engine.table.impl.remote.ConstructSnapshot.serializeAllTable(ConstructSnapshot.java:1514)
	at io.deephaven.engine.table.impl.remote.ConstructSnapshot.lambda$constructBackplaneSnapshotInPositionSpace$2(ConstructSnapshot.java:703)
	at io.deephaven.engine.table.impl.remote.ConstructSnapshot.callDataSnapshotFunction(ConstructSnapshot.java:1242)
	... 15 more
```

I don't have an exact reproducer for this, but here's the rough idea that gets to the structure of the ring table being fed from the blink table:
```python
from deephaven.table import Table
from deephaven.table_factory import ring_table
from typing import Union, List

def blink_tail_by(blink_table: Table, num_rows: int, by: Union[str, List[str]]) -> Table:
    """
    Transform blink_table to ring_table that keeps num_rows by <group>
    Workaround as suggested in https://github.com/deephaven/deephaven-core/issues/4309

    """
    return (
        blink_table.without_attributes(["BlinkTable"])
        .partitioned_agg_by(aggs=[], by=by, preserve_empty=True)
        .transform(lambda t: ring_table(t, num_rows))
        .merge()
    )


upstream = time_table("PT1s", blink_table=True).update(["I=nanosToSeconds(epochNanos(Timestamp))", "J=I%100", "K=(int) (I/100)", "Data=`1,2,3,4`.split(`,`)"]).ungroup()

downstream = blink_tail_by(upstream, 100, "K")
```

The upstream table is better defined as a table publisher (i.e. a blink table) with contents from two sorted tables, merged into one table and published.

--- a/server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java
+++ b/server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java
@@ -2126,7 +2126,7 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
 
     private class SnapshotControl implements ConstructSnapshot.SnapshotControl {
         long capturedLastUpdateClockStep;
-        long step = -1;
+        long resultValidStep = -1;
         final List<Subscription> snapshotSubscriptions;
 
         SnapshotControl(final List<Subscription> snapshotSubscriptions) {
@@ -2145,14 +2145,14 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
             final LogicalClock.State beforeState = LogicalClock.getState(beforeClockValue);
             final long beforeStep = LogicalClock.getStep(beforeClockValue);
             if (beforeState == LogicalClock.State.Idle) {
-                step = beforeStep;
+                resultValidStep = beforeStep;
                 return false;
             }
 
             final boolean notifiedOnThisStep = beforeStep == capturedLastUpdateClockStep;
             final boolean usePrevious = !notifiedOnThisStep;
 
-            step = notifiedOnThisStep ? beforeStep : beforeStep - 1;
+            resultValidStep = notifiedOnThisStep ? beforeStep : beforeStep - 1;
 
             if (log.isDebugEnabled()) {
                 log.debug().append(logPrefix)
@@ -2179,7 +2179,7 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
                 success = snapshotConsistent(afterClockValue, usedPreviousValues);
 
                 if (!success) {
-                    step = -1;
+                    resultValidStep = -1;
                 } else {
                     flipSnapshotStateForSubscriptions(snapshotSubscriptions);
                     finalizeSnapshotForSubscriptions(snapshotSubscriptions);
@@ -2191,7 +2191,7 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
             }
             if (log.isDebugEnabled()) {
                 log.debug().append(logPrefix)
-                        .append("success=").append(success).append(", step=").append(step).endl();
+                        .append("success=").append(success).append(", validStep=").append(resultValidStep).endl();
             }
             return success;
         }

--- a/server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java
+++ b/server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java
@@ -2145,19 +2145,19 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
             final LogicalClock.State beforeState = LogicalClock.getState(beforeClockValue);
             final long beforeStep = LogicalClock.getStep(beforeClockValue);
             if (beforeState == LogicalClock.State.Idle) {
-                this.step = beforeStep;
+                step = beforeStep;
                 return false;
             }
 
-            final boolean notifiedOnThisStep = step == capturedLastIndexClockStep;
+            final boolean notifiedOnThisStep = beforeStep == capturedLastIndexClockStep;
             final boolean usePrevious = !notifiedOnThisStep;
 
-            this.step = notifiedOnThisStep ? step : step - 1;
+            step = notifiedOnThisStep ? beforeStep : beforeStep - 1;
 
             if (log.isDebugEnabled()) {
                 log.debug().append(logPrefix)
-                        .append("previousValuesAllowed usePrevious=").append(usePrevious)
-                        .append(", step=").append(step).append(", validStep=").append(this.step).endl();
+                        .append("usePreviousValues usePrevious=").append(usePrevious)
+                        .append(", beforeStep=").append(beforeStep).append(", validStep=").append(step).endl();
             }
 
             return usePrevious;

--- a/server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java
+++ b/server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java
@@ -244,7 +244,7 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
      * This is the last step on which the UG-synced RowSet was updated. This is used only for consistency checking
      * between our initial creation and subsequent updates.
      */
-    private long lastIndexClockStep = 0;
+    private long lastUpdateClockStep = 0;
 
     private Throwable pendingError = null;
     private final List<Delta> pendingDeltas = new ArrayList<>();
@@ -603,8 +603,8 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
         @Override
         public void onUpdate(final TableUpdate upstream) {
             synchronized (BarrageMessageProducer.this) {
-                if (lastIndexClockStep >= parent.getUpdateGraph().clock().currentStep()) {
-                    throw new IllegalStateException(logPrefix + "lastIndexClockStep=" + lastIndexClockStep
+                if (lastUpdateClockStep >= parent.getUpdateGraph().clock().currentStep()) {
+                    throw new IllegalStateException(logPrefix + "lastUpdateClockStep=" + lastUpdateClockStep
                             + " >= notification on "
                             + parent.getUpdateGraph().clock().currentStep());
                 }
@@ -619,11 +619,11 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
                 parentTableSize = parent.size();
 
                 // mark when the last indices are from, so that terminal notifications can make use of them if required
-                lastIndexClockStep = parent.getUpdateGraph().clock().currentStep();
+                lastUpdateClockStep = parent.getUpdateGraph().clock().currentStep();
                 if (log.isDebugEnabled()) {
                     try (final RowSet prevRowSet = parent.getRowSet().copyPrev()) {
                         log.debug().append(logPrefix)
-                                .append("lastIndexClockStep=").append(lastIndexClockStep)
+                                .append("lastUpdateClockStep=").append(lastUpdateClockStep)
                                 .append(", upstream=").append(upstream).append(", shouldEnqueueDelta=")
                                 .append(shouldEnqueueDelta)
                                 .append(", rowSet=").append(parent.getRowSet()).append(", prevRowSet=")
@@ -2120,12 +2120,12 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
         postSnapshotColumns.clear();
     }
 
-    private synchronized long getLastIndexClockStep() {
-        return lastIndexClockStep;
+    private synchronized long getLastUpdateClockStep() {
+        return lastUpdateClockStep;
     }
 
     private class SnapshotControl implements ConstructSnapshot.SnapshotControl {
-        long capturedLastIndexClockStep;
+        long capturedLastUpdateClockStep;
         long step = -1;
         final List<Subscription> snapshotSubscriptions;
 
@@ -2140,7 +2140,7 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
                 return false;
             }
 
-            capturedLastIndexClockStep = getLastIndexClockStep();
+            capturedLastUpdateClockStep = getLastUpdateClockStep();
 
             final LogicalClock.State beforeState = LogicalClock.getState(beforeClockValue);
             final long beforeStep = LogicalClock.getStep(beforeClockValue);
@@ -2149,15 +2149,16 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
                 return false;
             }
 
-            final boolean notifiedOnThisStep = beforeStep == capturedLastIndexClockStep;
+            final boolean notifiedOnThisStep = beforeStep == capturedLastUpdateClockStep;
             final boolean usePrevious = !notifiedOnThisStep;
 
             step = notifiedOnThisStep ? beforeStep : beforeStep - 1;
 
             if (log.isDebugEnabled()) {
                 log.debug().append(logPrefix)
-                        .append("usePreviousValues usePrevious=").append(usePrevious)
-                        .append(", beforeStep=").append(beforeStep).append(", validStep=").append(step).endl();
+                        .append("usePreviousValues: usePrevious=").append(usePrevious)
+                        .append(", beforeStep=").append(beforeStep)
+                        .append(", lastUpdateStep=").append(capturedLastUpdateClockStep).endl();
             }
 
             return usePrevious;
@@ -2168,7 +2169,7 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
             if (!parentIsRefreshing) {
                 return true;
             }
-            return capturedLastIndexClockStep == getLastIndexClockStep();
+            return capturedLastUpdateClockStep == getLastUpdateClockStep();
         }
 
         @Override
@@ -2336,6 +2337,6 @@ public class BarrageMessageProducer<MessageView> extends LivenessArtifact
 
     @Override
     public synchronized void setLastNotificationStep(final long lastNotificationStep) {
-        lastIndexClockStep = Math.max(lastNotificationStep, lastIndexClockStep);
+        lastUpdateClockStep = Math.max(lastNotificationStep, lastUpdateClockStep);
     }
 }

